### PR TITLE
[feature](scan) support split scanners by size of segment

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

File: be/src/olap/parallel_scanner_builder.cpp [hunk 1]
```diff
@@ -32,13 +32,120 @@ template <typename ParentType>
...
+template <typename ParentType>
+Status ParallelScannerBuilder<ParentType>::_build_scanners_by_segment_size(
+        std::list<VScannerSPtr>& scanners) {
+    DCHECK_GE(_bytes_per_scanner, 0);
+
+    for (auto&& [tablet, version] : _tablets) {
+        DCHECK(_all_rowsets.contains(tablet->tablet_id()));
+        auto& rowsets = _all_rowsets[tablet->tablet_id()];
+
+        TabletReader::ReadSource reade_source_with_delete_info;
```
Comment (yiguolei) [inline be/src/olap/parallel_scanner_builder.cpp:54]: typo

File: be/src/olap/parallel_scanner_builder.cpp [hunk 2]
```diff
@@ -32,13 +32,120 @@ template <typename ParentType>
...
+            int segment_start = 0;
+            auto split = RowSetSplits(reader->clone());
+
+            for (size_t i = 0; i != segments_size.size(); ++i) {
+                const auto segment_size = segments_size[i];
+                RowRanges row_ranges;
+
+                row_ranges.add({0, std::numeric_limits<int64_t>::max()});
+                split.segment_row_ranges.emplace_back(std::move(row_ranges));
+
```
Comment (yiguolei) [inline be/src/olap/parallel_scanner_builder.cpp:92]: if segment1 = 31MB and segment2 = 31mb, then collectored is 62MB, it maybe too large.


I think maybe we should sort the segment by size and then we could use some algoriithm to make the split more well distributed between scanners.

**Comment by mrhhsg** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 3]
```diff
@@ -951,6 +955,14 @@ public class SessionVariable implements Serializable, Writable {
             varType = VariableAnnotation.EXPERIMENTAL, needForward = true)
     private long parallelScanMinRowsPerScanner = 16384; // 16K
 
+    @VariableMgr.VarAttr(name = PARALLEL_SCAN_SEGMENT_COUNT_FACTOR, fuzzy = true,
+                varType = VariableAnnotation.EXPERIMENTAL, needForward = true)
+    private long parallelScanSegmentCountFactor = 5;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:960]: default to 10

File: be/src/olap/parallel_scanner_builder.h [hunk 4]
```diff
@@ -77,11 +82,19 @@ class ParallelScannerBuilder {
...
+
+    std::map<RowsetId, std::vector<size_t>> _segments_size;
+
     std::map<RowsetId, SegmentCacheHandle> _segment_cache_handles;
 
     std::shared_ptr<RuntimeProfile> _scanner_profile;
     RuntimeState* _state;
     int64_t _limit_per_scanner;
+    bool _split_by_segment_size {false};
     bool _is_dup_mow_key;
```
Comment (yiguolei) [inline be/src/olap/parallel_scanner_builder.h:98]: and  also init other properties such as _is_dup_mow_key _limit_per_scanner..... should init all pointer to nullptr

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Pull request closed without merging**